src/solvers/smt2_incremental/smt_sorts.h

```diff
@@ -0,0 +1,91 @@
+// Author: Diffblue Ltd.
+
+/// \file
+/// Data structure for smt sorts.
+/// \note All classes derived from `smt_sortt`, must be listed in smt_sorts.def.
+
+#ifndef CPROVER_SOLVERS_SMT2_INCREMENTAL_SMT_SORTS_H
+#define CPROVER_SOLVERS_SMT2_INCREMENTAL_SMT_SORTS_H
+
+#include <util/irep.h>
+
+#include <type_traits>
+
+class smt_sort_const_downcast_visitort;
+
+class smt_sortt : protected irept
```

Comment:
Why define these separately when you could use `typet`?  The type system is essentially identical.  I can see a sort of value in having a `class smt_sortt : public typet` and a `smt_sortt to_smt_sort(typet &t)` function so you can use the C++ type safety to enforce conversion but this really seems like overkill.